craft_providers/multipass/multipass.py

```diff
@@ -386,3 +387,65 @@ def umount(self, *, mount: str) -> None:
                 brief=f"Failed to unmount {mount!r}.",
                 details=errors.details_from_called_process_error(error),
             ) from error
+
+    def wait_until_ready(
+        self, *, retry_wait: float = 0.25, timeout: Optional[float] = None
+    ):
+        """Wait until Multipass is ready (upon install/startup).
+
+        :param retry_wait: Time to sleep between retries.
+        :param timeout: Optional time.time() stamp to timeout after.
+        """
+        while timeout is None or time.time() < timeout:
+            multipass_version, multipassd_version = self.version()
+
+            if multipassd_version is not None:
+                return (multipass_version, multipassd_version)
+
+            time.sleep(retry_wait)
+
+        raise MultipassError(
+            brief="Timed out waiting for Multipass to become ready.",
+        )
+
+    def version(self):
+        """Get multipass and multipassd versions.
+
+        Split multipass version output should look like:
+
+            - ['multipass', '1.5.0']
+
+            - ['multipass', '1.5.0', 'multipassd', '1.5.0']
+
+            - ['multipass', '1.5.0+mac', 'multipassd', '1.5.0+mac']
+
+            - ['multipass', '1.5.0+win', 'multipassd', '1.5.0+win']
+
+        In some cases, Multipass may spew additional data, so protect against
+        it.  See: https://github.com/canonical/multipass/issues/2020
+
+        :returns: Tuple of parsed versions (multipass, multipassd).  multipassd
+                  may be None if Multipass is not yet ready.
+        """
+        try:
+            proc = self._run(["version"], capture_output=True, check=True, text=True)
```

Comment:
text=True, was not aware of that one, do we get decode errors if decoding fails?

Reply:
Explicitly added decoding failure logic